# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: gfx/thebes/SharedFontList.cpp
# Commit: ac4ab326e785
# Full Hash: ac4ab326e785a5ad42ce51b54749c2bcce394916
# Author: Jonathan Kew <jkew@mozilla.com>
# Date: 2024-02-19 21:22:25
# Description:
#   Bug 1879825 - Make Family::FindIndex fallible and safely handle failures. r=gfx-reviewers,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D201819
# ==============================================================================

diff -r e4fcd942953c -r ac4ab326e785 gfx/thebes/SharedFontList.cpp
--- a/gfx/thebes/SharedFontList.cpp	Mon Feb 19 12:24:19 2024 +0000
+++ b/gfx/thebes/SharedFontList.cpp	Mon Feb 19 12:37:33 2024 +0000
@@ -168,7 +168,11 @@
 void Face::SetCharacterMap(FontList* aList, gfxCharacterMap* aCharMap,
                            const Family* aFamily) {
   if (!XRE_IsParentProcess()) {
-    std::pair<uint32_t, bool> familyIndex = aFamily->FindIndex(aList);
+    Maybe<std::pair<uint32_t, bool>> familyIndex = aFamily->FindIndex(aList);
+    if (!familyIndex) {
+      NS_WARNING("Family index not found! Ignoring SetCharacterMap");
+      return;
+    }
     const auto* faces = aFamily->Faces(aList);
     uint32_t faceIndex = 0;
     while (faceIndex < aFamily->NumFaces()) {
@@ -177,14 +181,17 @@
       }
       ++faceIndex;
     }
-    MOZ_RELEASE_ASSERT(faceIndex < aFamily->NumFaces(), "Face ptr not found!");
+    if (faceIndex >= aFamily->NumFaces()) {
+      NS_WARNING("Face not found in family! Ignoring SetCharacterMap");
+      return;
+    }
     if (NS_IsMainThread()) {
       dom::ContentChild::GetSingleton()->SendSetCharacterMap(
-          aList->GetGeneration(), familyIndex.first, familyIndex.second,
+          aList->GetGeneration(), familyIndex->first, familyIndex->second,
           faceIndex, *aCharMap);
     } else {
       NS_DispatchToMainThread(new SetCharMapRunnable(
-          aList->GetGeneration(), familyIndex, faceIndex, aCharMap));
+          aList->GetGeneration(), familyIndex.value(), faceIndex, aCharMap));
     }
     return;
   }
@@ -629,16 +636,20 @@
   if (!XRE_IsParentProcess()) {
     // |this| could be a Family record in either the Families() or Aliases()
     // arrays; FindIndex will map it back to its index and which array.
-    std::pair<uint32_t, bool> index = FindIndex(aList);
+    Maybe<std::pair<uint32_t, bool>> index = FindIndex(aList);
+    if (!index) {
+      NS_WARNING("Family index not found! Ignoring SetupFamilyCharMap");
+      return;
+    }
     if (NS_IsMainThread()) {
       dom::ContentChild::GetSingleton()->SendSetupFamilyCharMap(
-          aList->GetGeneration(), index.first, index.second);
+          aList->GetGeneration(), index->first, index->second);
       return;
     }
     NS_DispatchToMainThread(NS_NewRunnableFunction(
         "SetupFamilyCharMap callback",
-        [gen = aList->GetGeneration(), idx = index.first,
-         alias = index.second] {
+        [gen = aList->GetGeneration(), idx = index->first,
+         alias = index->second] {
           dom::ContentChild::GetSingleton()->SendSetupFamilyCharMap(gen, idx,
                                                                     alias);
         }));
@@ -685,13 +696,13 @@
   }
 }
 
-std::pair<uint32_t, bool> Family::FindIndex(FontList* aList) const {
+Maybe<std::pair<uint32_t, bool>> Family::FindIndex(FontList* aList) const {
   const auto* start = aList->Families();
   const auto* end = start + aList->NumFamilies();
   if (this >= start && this < end) {
     uint32_t index = this - start;
     MOZ_RELEASE_ASSERT(start + index == this, "misaligned Family ptr!");
-    return std::pair(index, false);
+    return Some(std::pair(index, false));
   }
 
   start = aList->AliasFamilies();
@@ -699,10 +710,10 @@
   if (this >= start && this < end) {
     uint32_t index = this - start;
     MOZ_RELEASE_ASSERT(start + index == this, "misaligned AliasFamily ptr!");
-    return std::pair(index, true);
+    return Some(std::pair(index, true));
   }
 
-  MOZ_CRASH("invalid font-list Family ptr!");
+  return Nothing();
 }
 
 FontList::FontList(uint32_t aGeneration) {